### DISCUSSION TITLE
Review of `nawqa_wqp` example pipeline

### DISCUSSION THREAD

**Discussion by padilla410** [USGS-R/ds-pipelines-targets-example-wqp#8] [Ideas]
[`nawqa_wqp`](https://github.com/USGS-R/nawqa_wqp) repo contains a `scipiper` pipeline that grabs data from WQP and NWIS. Because of the focus of the `ds-pipelines-targets-example-wqp` repo this review will only focus on the portion of the WQP workflow.

Overall, `nawqa_wqp` repo is well organized and well documented. I was not able to run the complete pipeline because I ran into a few errors and timeout issues that did not seem worth the effort to troubleshoot. Because of our focus on WQP-related data pulls, [`1_wqpdata.yml`](https://github.com/USGS-R/nawqa_wqp/blob/master/1_wqpdata.yml) is the main file of interest in this review. In `scipiper`, these `yml` files are similar to the `_targets.R` file - they contain the recipe for each major step. Many of the design decisions in `1_wqpdata.yml` feel similar to the [`strategies for defining targets in data pipelines`](https://github.com/padilla410/ds-pipelines-targets-2/issues/4) portion of the`ds-pipelines-targets-2` training series because of the use of configuration files and the division of the data pull based on parameter and spatial extent.

### Workflow Overview
The WQP portion of the pipeline (`1_wqpdata`) completes the following when querying WQP data:

* Loads data configuration information that includes the following parameters:
  * Temporal extent: 2010-2015
  * Spatial extent: all 50 states and multiple territories
  * Defines parameters of interest - there is some basic grouping of parameters into parameter families (this could be considered a very light, but necessary, munging step considering the wide variety of similar parameter names in WQP)
  * Sets the max pull size to 50,000 records
* Creates inventory that summarizes data by parameter family, state, and pull size and places it under `out/inventory`
* Pulls data
  * Grabs data and site information for each pull and puts it in `out/data`
    * Data pull partitions are created using state, parameter family, and a max pull # specified in the configuration
    * Merges site info with data (could be considered another light munge)
  * Logs the date and time each file was pulled in `log` subfolder

I was unable to generate the task plan that completed the data pull step. I took a reasonable guess based reading `wqp_pull.R`.

There are a few things that this repo specifically does not do:

* It does not run parallel queries on WQP - [link](https://github.com/USGS-R/nawqa_wqp/blob/970d12fae76a23a3ce692d95d4cd8e8daa03707a/1_wqpdata/src/wqp_pull.R#L3-L8)
* it does not separate the local data download and the post to google drive - [link](https://github.com/USGS-R/nawqa_wqp/blob/970d12fae76a23a3ce692d95d4cd8e8daa03707a/1_wqpdata.yml#L67-L70)

A few quirky issues I found:
* There is a [`wqp_munge.R`](https://github.com/USGS-R/nawqa_wqp/blob/main/1_wqpdata/src/wqp_munge.R) that looks like it can create munge task plan, but I did not see these functions called anywhere else in the repo
* There is a [`4_data_gap_analysis`](https://github.com/USGS-R/nawqa_wqp/tree/main/4_gap_analysis) piece of the pipeline, but it looks like this portion of the repo exists outside of the `scipiper` pipeline. It’s a series of scripts and batch files.

### `Targets` training concepts covered

* Dynamic branching - uses the `scipiper` task table concept instead of `targets` dynamic branching
* File organization - uses `1_wqpdata `, `3_wtemp`, `4_gap_analysis` 

### Missing pieces, places to improve, and general comments:

* It is unclear why the wqp pull size is set to [50,000](https://github.com/USGS-R/nawqa_wqp/blob/970d12fae76a23a3ce692d95d4cd8e8daa03707a/1_wqpdata/cfg/wqp_pull.yml#L2). When we partition our pulls we should include a brief discussion on how to choose a value.
* In general, I love the inclusion of a data gap analysis. Perhaps this is something that we could consider as a secondary goal, after the data pull and light munge that we previously discussed.
* Tons of great commentary in the yml files and R function files, but there is no comprehensive `README` at the top of the repo - having one could add a little clarity. Additionally, I would say that the functions are about half documented – text explaining what they do, but I also like to see argument definitions and some info on what the output is.

### A few thoughts on our pipeline:
* Based on what I've seen, I think it would be reasonable to query our data based on parameter family and HUC (size TBD). Branching this way could prevent us from having to re-query all portions of the data in the future.

**Comment by lekoenig** [comment]
Thanks, @padilla410 - this is a really helpful summary! I'm guessing that some of the elements in the `nawqa_wqp` may be more advanced than we'll plan to implement in this pipeline. For example, the way `nawqa_wqp` chunks up and partitions the data pulls is awesome 🤩 - but do we expect to partition data pulls in this pipeline (other than branching over HUC's or CharacteristicNames)?


> Creates inventory that summarizes data by parameter family, state, and pull size and places it under out/inventory

I ❤️  this idea for our pipeline because I think this sort of quick summary is something many users would want. I also think we could use a table like that to illustrate the beauty of a reproducible pipeline: e.g. change your `endDate`, run `tar_make()`, and look! the inventory now captures X more samples!
